lib/engine.ml

```diff
@@ -288,21 +289,31 @@ let rec separate_records : string -> ((tls_hdr * string) list * string, failure)
     Error (`Fatal e)
 
 let encrypt_records encryptor version records =
-  let rec split = function
-    | [] -> []
-    | (t1, a) :: xs when String.length a >= 1 lsl 14 ->
-      let fst, snd = split_str a (1 lsl 14) in
-      (t1, fst) :: split ((t1, snd) :: xs)
-    | x::xs -> x :: split xs
-
-  and crypt st = function
-    | []            -> (st, [])
-    | (ty, buf)::rs ->
-        let (st, ty, enc) = encrypt version st ty buf in
-        let (st, encs) = crypt st rs in
-        (st, (ty, enc) :: encs)
+  let rec crypt st acc = function
+    | [] -> st, List.rev acc
+    | (ty, buf) :: rest ->
+      let bufl = String.length buf in
+      let rec doit st acc off =
+        if bufl - off >= 1 lsl 14 then
+          let len = 1 lsl 14 in
+          let st, ty, buf = encrypt version st ty buf off len in
+          doit st ((ty, buf) :: acc) (off + len)
+        else
+          let st, ty, buf = encrypt version st ty buf off (bufl - off) in
+          st, (ty, buf) :: acc
+      in
+      let st, res =
+        if String.length buf >= 1 lsl 14 then
+          doit st [] 0
+        else
+          let st, ty, buf =
+            encrypt version st ty buf 0 (String.length buf)
+          in
+          st, [ ty, buf ]
```

Comment:
Isn't this the same as just calling `doit st [] 0`?

Reply:
yes indeed, fixed in 8ff2cb3